rust/otap-dataflow/crates/admin/src/telemetry.rs

```diff
@@ -0,0 +1,866 @@
+        let mut attrs: Vec<(&String, &AttributeValue)> = g.attributes.iter().collect();
+        attrs.sort_by(|a, b| a.0.cmp(b.0));
+        for (k, v) in attrs {
+            if !base_labels.is_empty() {
+                base_labels.push(',');
+            }
+            let _ = write!(
+                &mut base_labels,
+                "{}=\"{}\"",
+                sanitize_prom_label_key(k),
+                escape_prom_label_value(&v.to_string_value())
+            );
+        }
+
+        // Emit metrics for this group
+        for field in g.brief.metrics.iter() {
+            if let Some(value) = g.metrics.get(field.name) {
+                let metric_name = sanitize_prom_metric_name(field.name);
+                if seen.insert(metric_name.clone()) {
+                    if !field.brief.is_empty() {
+                        let _ = writeln!(
+                            &mut out,
+                            "# HELP {} {}",
+                            metric_name,
+                            escape_prom_help(field.brief)
+                        );
+                    }
+                    let prom_type = match field.instrument {
+                        Instrument::Counter => "counter",
+                        Instrument::UpDownCounter => "gauge",
+                        Instrument::Gauge => "gauge",
+                        Instrument::Histogram => "histogram",
+                    };
+                    let _ = writeln!(&mut out, "# TYPE {} {}", metric_name, prom_type);
+                }
+                if base_labels.is_empty() {
+                    let _ = writeln!(&mut out, "{} {}{}", metric_name, value, ts_suffix);
+                } else {
+                    let _ = writeln!(
+                        &mut out,
+                        "{}{{{}}} {}{}",
+                        metric_name, base_labels, value, ts_suffix
+                    );
+                }
+            }
+        }
+    }
+
+    out
+}
+
+/// Collects a snapshot of current metrics without resetting them.
+fn collect_metrics_snapshot(registry: &MetricsRegistryHandle) -> Vec<MetricSetWithMetadata> {
+    let mut metric_sets = Vec::new();
+
+    registry.visit_current_metrics(|descriptor, attributes, metrics_iter| {
+        let mut metrics = Vec::new();
+
+        for (field, value) in metrics_iter {
+            metrics.push(MetricDataPointWithMetadata {
+                metadata: *field,
+                value,
+            });
+        }
+
+        if !metrics.is_empty() {
+            // Convert attributes to HashMap using the iterator
+            let mut attrs_map = HashMap::new();
+            for (key, value) in attributes.iter_attributes() {
+                let _ = attrs_map.insert(key.to_string(), value.clone());
+            }
+
+            metric_sets.push(MetricSetWithMetadata {
+                name: descriptor.name.to_owned(),
+                brief: "".to_string(), // MetricsDescriptor doesn't have description field
```

Comment:
nit: Using `String::new()` might be slightly better.

```suggestion
                brief: String::new(), // MetricsDescriptor doesn't have description field
```

Reply:
I applied this suggestion in several locations